File: backend/src/test/java/kh/lifelink/api/request/CambodianPhoneTest.java
package kh.lifelink.api.request;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class CambodianPhoneTest {

    @ParameterizedTest
    @ValueSource(
            strings = {
                "012345678",
                "012 345 678",
                "012-345-678",
                "+855 12 345 678",
                "+85512345678",
                "855012345678",
                "00855 12345678"
            })
    void acceptsASixDigitPrefixHoweverTyped(String input) {
        assertThat(CambodianPhone.normalize(input)).contains("+85512345678");
    }

    @ParameterizedTest
    @ValueSource(strings = {"097 123 4567", "0971234567", "+855 97 123 4567"})
    void acceptsAStarredPrefixWithSevenDigits(String input) {
        assertThat(CambodianPhone.normalize(input)).contains("+855971234567");
    }

    @ParameterizedTest
    @ValueSource(
            strings = {
                "",
                "01",
                "097 123 456", // starred prefix, six digits
                "0123456789", // six-digit prefix, seven digits
                "023 123 456", // Phnom Penh landline
                "013 123 456", // not allocated to mobile
                "012 ABC 678"
            })
    void rejectsAnythingElse(String input) {
        assertThat(CambodianPhone.normalize(input)).isEmpty();
    }
}

File: mobile/test/cambodian_phone_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lifelink_kh/src/core/phone/cambodian_phone.dart';

void main() {
    group('CambodianPhone.normalize', () {
        test('accepts a six-digit prefix with six subscriber digits, however typed', () {
            for (final input in [
                '012345678',
                '012 345 678',
                '012-345-678',
                '+855 12 345 678',
                '+85512345678',
                '855012345678',
                '00855 12345678',
            ]) {
                expect(CambodianPhone.normalize(input), '+85512345678', reason: input);
            }
        });

        test('accepts a starred prefix only with seven subscriber digits', () {
            expect(CambodianPhone.normalize('097 123 4567'), '+855971234567');
            expect(CambodianPhone.normalize('097 123 456'), isNull);
            expect(CambodianPhone.normalize('0181234567'), '+855181234567');
        });

        test('rejects a six-digit prefix with seven subscriber digits', () {
            expect(CambodianPhone.normalize('0123456789'), isNull);
        });

        test('rejects prefixes TRC has not allocated to mobile', () {
            expect(CambodianPhone.normalize('023 123 456'), isNull); // Phnom Penh landline
            expect(CambodianPhone.normalize('013 123 456'), isNull);
        });

        test('rejects letters, empties and fragments', () {
            expect(CambodianPhone.normalize(''), isNull);
            expect(CambodianPhone.normalize('01'), isNull);
            expect(CambodianPhone.normalize('012 ABC 678'), isNull);
        });
    });
}
